Assets/MRTK/SDK/Experimental/Dock/Dockable.cs

```diff
@@ -0,0 +1,335 @@
+    public class Dockable : MonoBehaviour
+    {
+        [Experimental]
+        [SerializeField, ReadOnly]
+        [Tooltip("Current state of this dockable in regards to a dock.")]
+        private DockingState dockingState = DockingState.Undocked;
+
+        [SerializeField]
+        [Tooltip("Time to animate any move/scale into or out of the dock.")]
+        private float moveLerpTime = 0.1f;
+
+        [SerializeField]
+        [Tooltip("Time to animate an element when it's following the dock (use 0 for tight attachment)")]
+        private float moveLerpTimeWhenDocked = 0.05f;
+
+        /// <summary>
+        /// True if this object can currently be docked, false otherwise.
+        /// </summary>
+        public bool CanDock => dockingState == DockingState.Undocked || dockingState == DockingState.Undocking;
+
+        /// <summary>
+        /// True if this object can currently be undocked, false otherwise.
+        /// </summary>
+        public bool CanUndock => dockingState == DockingState.Docked;
+
+        // Constants
+        private static readonly float distanceTolerance = 0.01f; // in meters
+        private static readonly float angleTolerance = 3.0f; // in degrees
+        private static readonly float scaleTolerance = 0.01f; // in percentage
+
+        private DockPosition dockedPosition = null;
+        private Vector3 dockedPositionScale = Vector3.one;
+
+        private HashSet<DockPosition> overlappingPositions = new HashSet<DockPosition>();
+        private Vector3 originalScale = Vector3.one;
+        private bool isDragging = false;
+        private ObjectManipulator objectManipulator;
+        private ManipulationHandler manipulationHandler;
+
+        /// <summary>
+        /// Subscribes to manipulation events.
+        /// </summary>
+        private void Start()
+        {
+            objectManipulator = gameObject.GetComponent<ObjectManipulator>();
+            if (objectManipulator != null)
+            {
+                objectManipulator.OnManipulationStarted.AddListener(OnManipulationStarted);
+                objectManipulator.OnManipulationEnded.AddListener(OnManipulationEnded);
+            }
+            else
+            {
+                manipulationHandler = gameObject.GetComponent<ManipulationHandler>();
+                if (manipulationHandler != null)
+                {
+                    manipulationHandler.OnManipulationStarted.AddListener(OnManipulationStarted);
+                    manipulationHandler.OnManipulationEnded.AddListener(OnManipulationEnded);
+                }
+            }
+
+            Assert.IsTrue(objectManipulator != null || manipulationHandler != null,
+                "A Dockable object must have either an ObjectManipulator or a ManipulationHandler component.");
+        }
+
+        /// <summary>
+        /// Unsubscribes from manipulation events.
+        /// </summary>
+        private void OnDestroy()
```

Comment:
Moving the initialization from `Start` to `OnEnable`, I'd move this to `OnDisable` to prevent the component registering multiple times.